server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -191,9 +191,47 @@ public static <Response extends ReplicationResponse & WriteResponse> ActionListe
 
     @Override
     protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+        final long startTime = relativeTime();
+        final ClusterState initialState = clusterService.state();
+        final ClusterBlockException blockException = initialState.blocks().globalBlockedException(ClusterBlockLevel.WRITE);
+        if (blockException != null) {
+            if (false == blockException.retryable()) {
+                listener.onFailure(blockException);
+                return;
+            }
+            logger.trace("cluster is blocked, waiting for it to recover", blockException);
+            final ClusterStateObserver clusterStateObserver = new ClusterStateObserver(
+                initialState,
+                clusterService,
+                bulkRequest.timeout(),
+                logger,
+                threadPool.getThreadContext()
+            );
+            clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    doExecuteOnWriteThreadPool(task, bulkRequest, startTime, listener);
+                }
+
+                @Override
+                public void onClusterServiceClose() {
+                    listener.onFailure(new NodeClosedException(clusterService.localNode()));
+                }
+
+                @Override
+                public void onTimeout(TimeValue timeout) {
+                    listener.onFailure(blockException);
+                }
+            }, newState -> false == newState.blocks().hasGlobalBlockWithLevel(ClusterBlockLevel.WRITE));
+        } else {
+            doExecuteOnWriteThreadPool(task, bulkRequest, startTime, listener);
+        }
+    }
+
+    private void doExecuteOnWriteThreadPool(Task task, BulkRequest bulkRequest, long startTime, ActionListener<BulkResponse> listener) {
```

Comment:
Naming nit: I interpreted this name as indicating that it's _called_ on the write threadpool. Maybe something like `forkAndExecute` would be clearer?

```suggestion
    private void forkAndExecute(Task task, BulkRequest bulkRequest, long startTime, ActionListener<BulkResponse> listener) {
```